### PR TITLE
Sarama logging via configmap

### PR DESCRIPTION
**Pull request by eric-sap**
## Proposed Changes

- 🎁  Enhance the config-eventing-kafka configmap by adding "enableSaramaLogging" fields to the dispatcher, receiver, and controller sections.  These allow the sarama.Logger to be configured dynamically without requiring a recompile and redeploy

### CODE REVIEW

File: pkg/channel/distributed/controller/kafkachannel/reconciler.go [hunk 1]
```diff
@@ -218,12 +218,21 @@ func (r *Reconciler) configMapObserver(configMap *corev1.ConfigMap) {
...
-	// section, we currently do not do anything proactive based on configuration changes to those items.
-	// The only component in the controller that uses any of the fields after startup currently
-	// is the AdminClient, which simply uses the r.saramaConfig set here whenever necessary.
-	// This means that calling env.GetEnvironment is not necessary now.  If
-	// those settings are needed in the future, the environment will also need to be re-parsed here.
+	// Enable Sarama Logging If Specified In ConfigMap
+	if ekConfig, err := kafkasarama.LoadEventingKafkaSettings(configMap); err == nil && ekConfig != nil {
+		kafkasarama.EnableSaramaLogging(ekConfig.Kafka.EnableSaramaLogging)
+		r.logger.Debug(fmt.Sprintf("Setting Sarama Logging to %v", ekConfig.Kafka.EnableSaramaLogging))
+	} else {
```
Comment (travis-minke-sap) [inline pkg/channel/distributed/controller/kafkachannel/reconciler.go:225]: let's use zap fields instead fo Sprintf...

r.logger.Debug("Setting Sarama Logging", zap.Bool("Enabled", ekConfig.Kafka.EnableSaramaLogging))

File: pkg/channel/distributed/dispatcher/dispatcher/dispatcher.go [hunk 2]
```diff
@@ -284,6 +284,14 @@ func (d *DispatcherImpl) ConfigChanged(configMap *v1.ConfigMap) Dispatcher {
 		// Some of the current config settings may not be overridden by the configmap (username, password, etc.)
 		kafkasarama.UpdateSaramaConfig(newConfig, d.SaramaConfig.ClientID, d.SaramaConfig.Net.SASL.User, d.SaramaConfig.Net.SASL.Password)
 
+		// Enable Sarama Logging If Specified In ConfigMap
+		if ekConfig, err := kafkasarama.LoadEventingKafkaSettings(configMap); err == nil && ekConfig != nil {
+			kafkasarama.EnableSaramaLogging(ekConfig.Kafka.EnableSaramaLogging)
+			d.Logger.Debug(fmt.Sprintf("Setting Sarama Logging to %v", ekConfig.Kafka.EnableSaramaLogging))
+		} else {
```
Comment (travis-minke-sap) [inline pkg/channel/distributed/dispatcher/dispatcher/dispatcher.go:291]: same sprintf comment as in reconciler.go

File: pkg/channel/distributed/receiver/producer/producer.go [hunk 3]
```diff
@@ -208,6 +209,14 @@ func (p *Producer) ConfigChanged(configMap *v1.ConfigMap) *Producer {
 		// Some of the current config settings may not be overridden by the configmap (username, password, etc.)
 		kafkasarama.UpdateSaramaConfig(newConfig, p.configuration.ClientID, p.configuration.Net.SASL.User, p.configuration.Net.SASL.Password)
 
+		// Enable Sarama Logging If Specified In ConfigMap
+		if ekConfig, err := kafkasarama.LoadEventingKafkaSettings(configMap); err == nil && ekConfig != nil {
+			kafkasarama.EnableSaramaLogging(ekConfig.Kafka.EnableSaramaLogging)
+			p.logger.Debug(fmt.Sprintf("Setting Sarama Logging to %v", ekConfig.Kafka.EnableSaramaLogging))
+		} else {
```
Comment (travis-minke-sap) [inline pkg/channel/distributed/receiver/producer/producer.go:216]: same formatting change

File: pkg/channel/distributed/common/kafka/sarama/sarama.go [hunk 4]
```diff
@@ -296,21 +301,29 @@ func LoadSettings(ctx context.Context) (*sarama.Config, *commonconfig.EventingKa
...
+
+	return saramaConfig, eventingKafkaConfig, err
+}
+
+func LoadEventingKafkaSettings(configMap *corev1.ConfigMap) (*commonconfig.EventingKafkaConfig, error) {
 	// Validate The ConfigMap Data
 	if configMap.Data == nil {
-		return nil, nil, fmt.Errorf("attempted to load configuration from empty configmap")
+		return nil, fmt.Errorf("attempted to load configuration from empty configmap")
 	}
```
Comment (travis-minke-sap) [inline pkg/channel/distributed/common/kafka/sarama/sarama.go:319]: check `configMap` is nill as well (before configMap.Data check) ?

File: pkg/channel/distributed/common/kafka/sarama/sarama_test.go [hunk 5]
```diff
@@ -164,10 +165,12 @@ func TestEnableSaramaLogging(t *testing.T) {
...
 
 	// Perform The Test
-	EnableSaramaLogging()
+	EnableSaramaLogging(true)
 
 	// Verify Results (Not Much Is Possible)
 	sarama.Logger.Print("TestMessage")
+
+	EnableSaramaLogging(false)
 }
```
Comment (travis-minke-sap) [inline pkg/channel/distributed/common/kafka/sarama/sarama_test.go:175]: print again after disabling? (i know it's just visual verification but better than nothing ; )

Comment (eric-sap) [reply pkg/channel/distributed/common/kafka/sarama/sarama_test.go:178]: done

File: pkg/channel/distributed/common/kafka/sarama/sarama.go (hunk 4 above)
Comment (eric-sap) [reply pkg/channel/distributed/common/kafka/sarama/sarama.go:319]: good idea, will do

File: pkg/channel/distributed/controller/kafkachannel/reconciler.go (hunk 1 above)
Comment (eric-sap) [reply pkg/channel/distributed/controller/kafkachannel/reconciler.go:225]: Will do; ReviewDog suggested as much as well

File: pkg/channel/distributed/dispatcher/dispatcher/dispatcher.go (hunk 2 above)
Comment (eric-sap) [reply pkg/channel/distributed/dispatcher/dispatcher/dispatcher.go:291]: Will do; ReviewDog suggested as much as well

File: pkg/channel/distributed/receiver/producer/producer.go (hunk 3 above)
Comment (eric-sap) [reply pkg/channel/distributed/receiver/producer/producer.go:215]: Will do; ReviewDog suggested as much as well

**Pull request merged**